simple_network_sim/mixing_matrix.py

```diff
@@ -0,0 +1,132 @@
+                    raise Exception(f'Invalid age range specified: "{a}"')
+        else:
+            self._lower = int(a)
+            self._upper = int(b)
+
+    def contains(self, age):
+        """Returns true if age is inside this age range."""
+        if age < self._lower:
+            return False
+        if self._upper == -1:
+            return True
+        return age < self._upper
+
+    def __str__(self):
+        if self._upper == -1:
+            return f"{self._lower}+"
+        return f"[{self._lower},{self._upper})"
+
+    def __eq__(self, other):
+        return self._lower == other._lower and self._upper == other._upper
+
+    def __neq__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return 1000*self._lower + self._upper  # We just need any hashing method
+
+
+class MixingRow:
+    """A mixing row is one row of a mixing table. This is a helper class. A row
+    represents a given population, and can return the expected number of
+    interactions (per day) a member of this population will have with some
+    other target.
+    """
+    def __init__(self, ages, interactions):
+        """Initialiser. Entries should be a dict mapping an AgeRange to an
+        expected number of interactions."""
+        self._entries = {}
+        for age, interact in zip(ages, interactions):
+            self._entries[age] = float(interact)
+
+    def __getitem__(self, age):
+        """Get the expected number of interactions (per day) that someone from
+        this MixingRow would have with someone with the given age, or age
+        range.
+        """
+        if isinstance(age, str) or isinstance(age, tuple):
+            return self._entries[AgeRange(age)]
+        for key, value in self._entries.items():
+            if key.contains(age):
+                return value
+        raise Exception(f'Could not find {age} in MixingRow')
+
+    def __str__(self):
+        return "[" + ", ".join(f"{str(key)}: {str(val)}"
+                               for key, val in self._entries.items()) + "]"
+
+
+class MixingMatrix:
+    """Can give the expected number of interactions per day a given person will
+    have with some other person, based on the ages of the two people, or given
+    age-ranges. A MixingMatrix will often be instantiated from a CSV file (see
+    initialiser). It can be used as follows:
+
+        mm = MixingMatrix("sample_input_file/sample_20200327_comix_social_contacts.sampleCSV")
+        print(mm[28][57])  # Prints the expected number of interactions a 28 year old
+                           # would have with a 57 year old in a day
+        print(mm["[30,40)"]["70+"])  # Prints the expected number of interactions someone in the
+                                     # age range [30-40) would have with someone aged 70 or older
+                                     # in any given day.
+        print(mm[(30,40)]["70+"])  # As above
+
+    """
+
+    def __init__(self, infile):
```

Comment:
Maybe it'd be interesting to raise an exception or print an error if there are any overlapping ranges. If someone messes up one of the ranges it could cause hard to detect bugs otherwise.

Reply:
Added a function to detect overlapping ranges, and adding some tests for these too.